### PR TITLE
[Relay] First order reverse mode automatic differentiation

### PR DESCRIPTION
**Pull request by MarisaKirisame**
- [x] implement helper function
- [x] move them to a header
- [x] implement reverse mode for operator
- [ ] test
- [x] doc and comment
- [x] GradWrt
- [ ] Defunctionalize the WList
- [x] Lift Relay Func into C++ Func for FuncFORAD

### CODE REVIEW

File: src/relay/pass/ad.cc [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+
+// This is implemented using wengert list. However, the list only exists at
+// compile time, and instead of holding runtime value of floats, it hold
+// code<float>. After the wengert list finish updating (local mutation at
+// compile time), we can use it to generate fully first order code. More
+// specifically, the work is done by transforming function literal (no free var)
+// by recursing, Real to a pair of poiner, pointing to the original code/the
+// wengert list. All others get partially evaluated to hold such value (so no
+// dynamic length list).
+struct FOReverseADError : Error {
```
Comment (yzhliu) [inline src/relay/pass/ad.cc]: qq: what does "FO" mean?

File: src/relay/pass/ad.cc [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+  explicit FOReverseADError(const std::string &msg)
+    : Error("ReverseADError: " + msg) {}
+};
+
+struct FORAD {
```
Comment (yzhliu) [inline src/relay/pass/ad.cc]: FORAD means FOReverse?

Comment (MarisaKirisame) [reply src/relay/pass/ad.cc]: First Order Reverse AD

File: src/relay/pass/ad.cc (hunk 1 above)
Comment (MarisaKirisame) [reply src/relay/pass/ad.cc]: First Order
what does qq mean BTW?

**Comment by MarisaKirisame** [comment]
I had made an end to end test that pretty print the result. I look correct, but as we have no evaluate yet, we cant test it's semantic. The rest is just filling in more op.

File: tests/python/relay/test_ad.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+from tvm.relay.op import log, add, equal, subtract, concat, multiply
+from tvm.relay.ir_builder import IRBuilder, convert
+
+def test_simple_gradient():
+    b = IRBuilder()
+    # f x = x * x + x
+    ty = relay.TensorType(tvm.convert([]), "float32")
+    x = relay.Var("x")
+    f = relay.Function([relay.Param(x, ty)], ty, multiply(x, x), [])
+    ret = fo_with_gradient(b.env, f)
```
Comment (ZihengJiang) [inline tests/python/relay/test_ad.py]: Do we need to emphasize "first-order" in the front API? Can we take `gradient` directly?

Comment (MarisaKirisame) [reply tests/python/relay/test_ad.py]: we can, but after discussion with tianqi and jared, we think it is best to supply a family of ad algorithm with different characteristic

**Comment by MarisaKirisame** [comment]
I had implement the reverse mode for all existing operator (plus a few that I add), except concat, and discrete one (equal).

**Comment by MarisaKirisame** [comment]
I had also document/refactor it. I think it is time for review.

File: src/relay/pass/let_list.h [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+   *
+   *  \param pv the var of the binding.
+   *
+   *  \param ty the type of the binding.
+   *
+   *  \param expr the value of the binding.
+   *
+   *  \return a PartialVar that hold the inserted expr.
+   */
+  PartialVar Let(const PartialVar& pv, const Type& ty, const Expr& expr) {
```
Comment (tqchen) [inline src/relay/pass/let_list.h]: Push

File: src/relay/pass/let_list.h [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+   */
+  PartialVar Let(const Expr& expr) {
+    return Let(IncompleteTypeNode::make(TypeParamNode::kType), expr);
+  }
+
+  /*! \brief wrap an expr around the LetList.
+   *
+   *  \return the wrapped expr.
+   */
+  Expr Plug(const Expr& expr) const {
```
Comment (tqchen) [inline src/relay/pass/let_list.h]: Get(const Expr& body)

File: src/relay/pass/let_list.h [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    Expr ret = expr;
+    for (auto rit = lets.rbegin(); rit != lets.rend(); rit++) {
```
Comment (tqchen) [inline src/relay/pass/let_list.h]: ++rit

File: src/relay/pass/let_list.h [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    return ret;
+  }
+
+  /*! \brief generate an LetList and wrap the result automatically.
+   *
+   *  \param f a function that generate the unwrapped Expr.
+   *
+   *  \return the wrapped Expr.
+   */
+  static Expr With(const std::function<Expr(LetList&)>& f) {
```
Comment (tqchen) [inline src/relay/pass/let_list.h]:
```
template<typename F>
static Expr With(F f) {
}
```

Comment (tqchen) [reply src/relay/pass/let_list.h]: https://github.com/dmlc/tvm/blob/master/include/tvm/runtime/packed_func.h#L57

Comment (tqchen) [reply src/relay/pass/let_list.h]: pass reference by pointer

**Comment by tqchen** [comment]
close this for now due to current inactive status, feel free to reopen later when there is new changes

**Pull request closed without merging**